honeybadgermpc/protocols/avid.py

```diff
@@ -0,0 +1,212 @@
+# coding=utf-8
+import logging
+import math
+import asyncio
+from honeybadgermpc.protocols.reliablebroadcast \
+    import encode, decode, merkle_tree, get_merkle_branch, merkle_verify
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.ERROR)
+# Uncomment this when you want logs from this file.
+# logger.setLevel(logging.NOTSET)
+
+
+class AVID:
+    # Class for Asynchronous Verifiable Information Dispersal
+    def __init__(self, n, f, leader, recv, send, input_size):
+        """ Initialize the class with parameters:
+        :param int n:  at least 3
+        :param int f: fault tolerance, ``N >= 3f + 1``
+        :param int leader: ``0 <= leader < N``
+        :param recv: :func:`recv()` blocks until a message is
+        recvd; message is of the form::
+            (i, (tag, ...)) = recv()
+        where ``tag`` is one of ``{"VAL", "ECHO", "READY"}``
+        :param send: sends (without blocking) a message to a designated
+        recipient ``send(i, (tag, ...))``
+        """
+        assert n >= 3 * f + 1
+        assert f >= 0
+        assert 0 <= leader < n
+
+        self._data = [None] * n
+        self.n = n
+        self.f = f
+        self.leader = leader
+        self.recv = recv
+        self.send = send
+        # size of input_list for disperse
+        self.input_size = input_size
+        # the response for retrieve
+        self.retrieval_queue = asyncio.Queue()
+        # A blocking queue: retrive will be sent until the ok state is reached
+        self.ok_queue = asyncio.Queue()
+
+    async def retrieve(self, sid, index):
+        """ Retrieve individual item
+        :param int index:  the index of retrieval
+        """
+        # we are able to retrieve things from AVID indivdually
+        assert 0 <= index < self.input_size
+
+        # send retrieve until ok state
+        assert await self.ok_queue.get()
+
+        # send retrieve to all the parties
+        for i in range(self.n):
+            self.send(i, (sid, "RETRIEVE", index))
```

Comment:
Define this strings as constants in a class instead of hard coding them (like an `AvidMessageType` class). Imagine, if you had a typo somewhere then you will end up not processing the message.

Reply:
Resolved.